pkg/storage/concurrency/concurrency_manager.go

```diff
@@ -0,0 +1,363 @@
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/google/btree"
+)
+
+// managerImpl implements the Manager interface.
+type managerImpl struct {
+	// Synchronizes conflicting in-flight requests.
+	lm latchManager
+	// Synchronizes conflicting in-progress transactions.
+	lt lockTable
+	// Ensures fair FIFO access to keys accessed by conflicting transactions.
+	lwq lockWaitQueueWaiter
+	// Waits for transaction completion and detects deadlocks.
+	twq txnWaitQueue
+	// The Store and Range that the manager is in charge of.
+	str Store
+	rng *roachpb.RangeDescriptor
+}
+
+// Store provides some parts of a Store without incurring a dependency.
+type Store interface {
+	// Components.
+	Clock() *hlc.Clock
+	Stopper() *stop.Stopper
+	DB() *client.DB
+	IntentResolver() *intentresolver.IntentResolver
+	// Knobs.
+	GetTxnWaitKnobs() txnwait.TestingKnobs
+	// Metrics.
+	GetTxnWaitMetrics() *txnwait.Metrics
+	GetSlowLatchGauge() *metric.Gauge
+}
+
+// NewManager creates a new concurrency Manager structure.
+func NewManager(store Store) Manager {
+	m := new(managerImpl)
+	// TODO(nvanbenschoten): move pkg/storage/spanlatch to a new
+	// pkg/storage/concurrency/latch package. Make it implement
+	// the latchManager interface directly, if possible.
+	m.lm = &latchManagerImpl{
+		m: spanlatch.Make(store.Stopper(), store.GetSlowLatchGauge()),
+	}
+	m.lt = &lockTableImpl{
+		qs: btree.New(16),
+	}
+	m.lwq = &lockWaitQueueWaiterImpl{}
+	// TODO(nvanbenschoten): move pkg/storage/txnwait to a new
+	// pkg/storage/concurrency/txnwait package.
+	m.twq = txnwait.NewQueue(store, m)
+	m.str = store
+	return m
+}
+
+// SequenceReq implements the Manager interface.
+func (m *managerImpl) SequenceReq(
+	ctx context.Context, prev *Guard, req Request,
+) (g *Guard, resp Response, err *Error) {
+	if !shouldAcquireLatches(req) {
+		return nil, nil, nil
+	}
+
+	// Ensure that we release the guard if we return a response or an error.
+	defer func() {
+		if g != nil && (resp != nil || err != nil) {
+			m.FinishReq(g)
+			g = nil
+		}
+	}()
+
+	// Provide the manager with an opportunity to intercept the request.
+	resp, err = m.maybeInterceptReq(ctx, req)
+	if resp != nil || err != nil {
+		return nil, resp, err
+	}
+
+	g = createOrReuseGuard(prev, req)
+	for {
+		// Acquire latches for the request. This synchronizes the request
+		// with all conflicting in-flight requests.
+		log.Event(ctx, "acquire latches")
+		g.lg, err = m.lm.Acquire(ctx, req)
+		if err != nil {
+			return g, nil, err
+		}
+
+		// Some requests don't want the wait on locks.
+		if !shouldWaitOnConflicts(req) {
+			return g, nil, nil
+		}
+
+		// Scan for conflicting locks.
+		log.Event(ctx, "scan for conflicting locks")
+		if wqgs := m.lt.ScanAndEnqueue(g.req, g); len(wqgs) > 0 {
+			g.wqgs = append(g.wqgs, wqgs...)
+		}
+
+		// Wait on each newly conflicting lock, if applicable.
+		if m.lwq.MustWaitOnAny(g.wqgs) {
```

Comment:
Could this be `wqgs` (which I know isn't in scope here)? Just trying to confirm my understanding that at the beginning of each iter in the `for` loop `g.wqgs` has been `WaitOn`'ed already and is only carried with the guard for .. other reasons.

Reply:
Originally that would have worked, but not anymore because in https://github.com/cockroachdb/cockroach/pull/43740 introduced the notion of "breaking" reservations. Essentially this means that waiters can lose their place at the front of a wait-queue while not holding latches. It is only safe to conclude that a waiter is at the front of all wait-queues if it is holding latches.

The reason for this is pretty subtle, but @sumeerbhola does a good job documenting it in that PR. Because we don't hold range locks while waiting in per-key wait-queues, we run into situations like the following. It's possible that two ranged writes _txn1_ and _txn2_ over keys [a, d) might sequence such that req _txn1_ initially acquires latches first. It finds a lock on key `c` and begins waiting. A different _txn3_ might then get sequenced and perform a point write on key `a`. When _txn2_ is sequenced it will enter the wait-queue for `a` (@ front) and `c` (behind _txn1_). When _txn1_ re-acquires latches and re-scans the lockTable, it will find the lock on `a` and enter its wait-queue. We don't want it getting behind _txn2_ in the queue or we'd cause a deadlock.

So, the solution Sumeer came up with to address this was to give each request a monotonically increasing sequence number when it first scans the lockTable. This creates a total order over requests. The order _does not_ tell us exactly what order requests will finish being sequenced by the concurrency manager. However, it does tell us how to order conflicting transactions in a given wait-queue. In other words, the policy for entering a wait-queue is not that requests automatically get in the back. Instead, requests scan from the back to find where they fit in the sequence-order list of reqs. This avoids the kinds of deadlocks discussed above.

However, to get it to work, we need to abandon the idea that once a request is at the front of a wait-queue, it stays there. Instead, requests much check all of their wait-queue entries on each iteration through the loop while holding latches. A request is only safe to proceed when it observes that it is at the front of all of its wait-queues during a single pass.

The correctness of this is subtle but I think it's mentioned in #43740. A consequence of it is that as a req with a low seq num, we need to be very careful about never breaking the reservation of another request whose latch set might be compatible with its, because latching is what's providing the mutual exclusion to atomically determine that a request holds a reservation for (is at the front of) all wait-queues that it's a part of. The fact that non-locking (`lock.None`) reads are never going to hold reservations in wait-queues makes this easier to think through.